### PR TITLE
created conda scrapper for conda-forge packages

### LINKED ISSUE #33: Create Conda Scraper for conda-forge packages
Scraping data from conda site for python packages.

### PR DESCRIPTION
**Pull request by siddheshghadi751**

### CODE REVIEW

File: score/conda.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+        if response.status_code == 200:
+            packages_name_list = response.json().keys()
+            logger.info("Total Packages : %s", len(packages_name_list))
+        else:
+            logger.error(
+                f"Failed to fetch URL: {packages_url} | status code: {response.status_code} | {response.text}"
+            )
+            packages_name_list = []
+    except Exception as e:
+        logger.error("Error fetching package names: %s", e)
```
Comment (srossross) [inline score/conda.py]: Why catch this exception? this will continue the process with no packages to build

File: score/conda.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+        logger.info("Fetching all packages from URL: %s", packages_url)
+        response = requests.get(packages_url)
+        if response.status_code == 200:
+            packages_name_list = response.json().keys()
+            logger.info("Total Packages : %s", len(packages_name_list))
+        else:
+            logger.error(
+                f"Failed to fetch URL: {packages_url} | status code: {response.status_code} | {response.text}"
+            )
+            packages_name_list = []
```
Comment (srossross) [inline score/conda.py]: see #17 use `response.raise_for_status()`

File: score/cli.py [hunk 3]
```diff
@@ -1,10 +1,14 @@
...
-from .logger import setup_logger
+
+import click
+
+from .conda import scrape_conda_packages
 from .data_retrieval.json_scraper import scrape_json
 from .data_retrieval.web_scraper import scrape_web
+from .logger import setup_logger
 
 OUTPUT_ROOT = Path(os.environ.get("OUTPUT_ROOT", "."))
```
Comment (srossross) [inline score/cli.py]: see #4 please move this to `cli.py` and pass it in as an argument

File: score/common.py [hunk 4]
```diff
@@ -0,0 +1,16 @@
+import pandas as pd
```
Comment (srossross) [inline score/common.py]: This function looks like it is just obfuscating the code and is unnecessary, lets just add the two lines inline

File: score/cli.py [hunk 5]
```diff
@@ -84,5 +88,21 @@ def scrape_pypi_web(start, end, output):
...
+    required=True,
+    help="Enter the starting letter or number to scrape (e.g., 'a' or '0').",
+)
+def conda(letter_to_scrape):
+    try:
+        subprocess.run(["python", "score/conda.py"], check=True)
+        scrape_conda_packages(letter_to_scrape)
+        click.echo("Scraping completed.")
+    except subprocess.CalledProcessError as e:
+        click.echo(f"Error: {e}")
```
Comment (srossross) [inline score/cli.py]: please don't catch and ignore exceptions. also why what would be causing a `subprocess.CalledProcessError`?

**Review by srossross — CHANGES_REQUESTED** [review]

File: score/conda.py (hunk 1 above)
Comment (siddheshghadi751) [reply score/conda.py]: removed this now. Its not needed.

File: requirements.txt [hunk 6]
```diff
@@ -1,6 +1,31 @@
...
+sniffio==1.3.1
+sortedcontainers==2.4.0
+soupsieve==2.5
+tqdm==4.66.5
+trio==0.26.0
+trio-websocket==0.11.1
+typing_extensions==4.12.2
+tzdata==2024.1
+urllib3==2.2.2
+websocket-client==1.8.0
```
Comment (srossross) [inline requirements.txt]: most of these packages look unused

Comment (siddheshghadi751) [reply requirements.txt]: I will remove them. Most packages got added due to installation of some used packages and their dependencies and when we do `pip freeze > requirements.txt` these extra packages got added

File: score/cli.py (hunk 3 above)
Comment (siddheshghadi751) [reply score/cli.py]: Its already in `cli.py`

File: requirements.txt [hunk 7]
```diff
@@ -1,6 +1,8 @@
+beautifulsoup4~=4.12.3
+bs4~=0.0.2
```
Comment (srossross) [inline requirements.txt]: what is this?

File: score/conda.py [hunk 8]
```diff
@@ -0,0 +1,137 @@
+import logging
+import os
+
+import pandas as pd
+import requests
+
+# Set up logger
+logging.basicConfig(level=logging.INFO)
+logger = logging.getLogger("conda")
+logger.addHandler(logging.FileHandler("conda.log"))
```
Comment (srossross) [inline score/conda.py]: Please remove this, it should be handled already by setup_logging in the cli.py

File: score/conda.py [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+    for package_name in list(packages_name_list):
+        package_data = get_package_data(package_name)
+        if package_data:
+            package_data_list.append(package_data)
+    if package_data_list:
+        df = pd.DataFrame(package_data_list)
+        df.to_parquet(output_dir, partition_cols=["initial_letter"])
+        logger.info("Data saved for package : %s", len(package_data_list))
+    else:
+        logger.error("Failed to save data for package letter : %s", letter_to_scrape)
```
Comment (srossross) [inline score/conda.py]: this should no longer be required
```suggestion

    df = pd.DataFrame(package_data_list)
    df.to_parquet(output_dir, partition_cols=["initial_letter"])
    logger.info("Data saved for package : %s", len(package_data_list))
```

File: requirements.txt [hunk 10]
```diff
@@ -1,6 +1,8 @@
...
 click~=8.1.7
+numpy~=2.0.1
```
Comment (srossross) [inline requirements.txt]: there are no numpy imports

**Review by srossross — CHANGES_REQUESTED** [review]
looks good. I've requested a few minor changes

File: score/conda.py (hunk 9 above)
Comment (siddheshghadi751) [reply score/conda.py]: This is updated now

File: requirements.txt (hunk 7 above)
Comment (siddheshghadi751) [reply requirements.txt]: beautifulsoup

File: requirements.txt (hunk 10 above)
Comment (siddheshghadi751) [reply requirements.txt]: removed them

**Comment by siddheshghadi751** [comment]
> looks good. I've requested a few minor changes

Done

File: score/conda.py [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+
+    Args:
+        letter (str): The starting letter to filter the package names.
+
+    Returns:
+        packages_name_list (list): A list of package names.
+
+    """
+    packages_name_list = list()
+    packages_url = "https://raw.githubusercontent.com/conda-forge/feedstock-outputs/single-file/feedstock-outputs.json"
```
Comment (srossross) [inline score/conda.py:23]: is this the correct conda url?

Comment (siddheshghadi751) [reply score/conda.py:23]: Yes, It is used in backend of conda-forge website.

**Pull request closed without merging**